src/FFmpegWriter.cpp

```diff
@@ -1525,11 +1525,23 @@ void FFmpegWriter::write_audio_packets(bool is_final) {
 			// Calculate total samples
 			total_frame_samples = samples_in_frame * channels_in_frame;
 
-			// Translate audio sample values back to 16 bit integers
-			for (int s = 0; s < total_frame_samples; s++, frame_position++)
-				// Translate sample value and copy into buffer
-				all_queued_samples[frame_position] = int(frame_samples_float[s] * (1 << 15));
-
+			// Translate audio sample values back to 16 bit integers with saturation
+			float valF;
+			int16_t conv;
+			const int16_t max16 = 32767;
+			const int16_t min16 = -32768;
+			for (int s = 0; s < total_frame_samples; s++, frame_position++) {
+				valF = frame_samples_float[s] * (1 << 15);
+				if (valF > max16)
+					conv = max16;
+				else if (valF < min16)
+					conv = min16;
+				else
+					conv = int(valF + 32768.5) - 32768; // +0.5 is for rounding
+
+				// Copy into buffer
+				all_queued_samples[frame_position] = conv;
+			}
```

Comment:
C++ defines `INT16_MAX` and `INT16_MIN` in the standard. (They're technically marked as optional, but if you `#include <cinttypes>` they should be there in any sane compiler.) So, if nothing else those can be used in place of `max16` and `min16`.

But you're absolutely right that it's a problem of bad limiting — it's really the `int()` there that's the killer. It allows the value to temporarily exceed `INT16_MIN`/`INT16_MAX` _during_ the computation, and then the overflow will produce bad values. Setting an `int16_t` variable to `int(1.1 * INT16_MAX)` doesn't produce `32767`, it produces `-29493`.

However, if we don't use `int()` for the math, the value will be limited correctly. This should work, without the temporary variables or manual limiting:
```c++
# At the top of the file
#include <cinttypes>

for (int s = 0; s < total_frame_samples; s++, frame_position++) {
    all_queued_samples[frame_position] = int16_t(frame_samples_float[s] * INT16_MAX));
}
```
(`INT16_MAX` is what `(1 << 15)` is manually computing.)

I was going to say that otherwise `valF` and/or `conv`should be declared _inside_ the loop for multithreading safety (despite the extra allocations), but this entire function is wrapped in `#pragma omp task firstprivate(is_final)` (and it contains plenty of other unsafe allocations already), so that's a moot point. Still, avoiding them is cleaner.

Reply:
Er, sorry, `#include <cinttypes>` is for the `std::fprintf` format constants. The `_MIN/MAX` typedefs should be included using `#include <cstdint>`. (Which cinttypes no doubt includes, itself, so it works either way. But it's more correct to use the cstdint header.) Above comment corrected.

Reply:
If you want to enforce the rounding direction, C++11 allows a [floating-point environment](https://en.cppreference.com/w/cpp/numeric/fenv/feround) to be established and configured. So, this would ensure that values `> N.5` are always rounded upwards to `N+1`:
```c++
#include <cfenv>
for (int s = 0; s < total_frame_samples; s++, frame_position++) {
    #pragma STDC FENV_ACCESS ON
    std::fesetround(FE_TONEAREST);
    all_queued_samples[frame_position] = int16_t(frame_samples_float[s] * INT16_MAX));
}
```
The scope will automatically be limited to the enclosing block (the inside of the for loop).

Reply:
\*sigh* Of course, it turns out that `#pragma` isn't supported, and `std::fesetround()` currently has no effect, in GCC or Clang. **#LAME.**

So, yeah, adding 0.5 seems to be the safest option for rounding.
```c++
for (int s = 0; s < total_frame_samples; s++, frame_position++) {
    all_queued_samples[frame_position] = int16_t(frame_samples_float[s] * INT16_MAX + 0.5));
}

Reply:
ARRGH. I'm wrong. It turns out the compiler is smart and fixes things, but only when it can do so trivially at compile time. e.g. this:
```c++
    int16_t out[4] = {
        int16_t(1.1 * (1 << 15)),
        int16_t(1.1 * INT16_MAX),
        int16_t(1.1 * (1 << 15) + 0.5),
        int16_t(1.1 * INT16_MAX + 0.5)
    };
    std::cout << std::setw(5) << 1.1 << ": "
              << std::setw(6) << out[0] << " "
              << std::setw(6) << out[1] << " "
              << std::setw(6) << out[2] << " "
              << std::setw(6) << out[3] << "\n";
    int16_t out2[4] = {
        int16_t(-1.1 * (1 << 15)),
        int16_t(-1.1 * INT16_MAX),
        int16_t(-1.1 * (1 << 15) + 0.5),
        int16_t(-1.1 * INT16_MAX + 0.5)
    };
    std::cout << std::setw(5) << -1.1 << ": "
              << std::setw(6) << out2[0] << " "
              << std::setw(6) << out2[1] << " "
              << std::setw(6) << out2[2] << " "
              << std::setw(6) << out2[3] << "\n";
```
Produces the correctly limited values:
```
  1.1:  32768  32768  32768  32768
 -1.1: -32768 -32768 -32768 -32768
```
But this:
```c++
void x(float f) {
    int16_t out[4] = {
        int16_t(f * (1 << 15)),
        int16_t(f * INT16_MAX),
        int16_t(f * (1 << 15) + 0.5),
        int16_t(f * INT16_MAX + 0.5)
    };
    std::cout << std::setw(5) << f << ": "
              << std::setw(6) << out[0] << " "
              << std::setw(6) << out[1] << " "
              << std::setw(6) << out[2] << " "
              << std::setw(6) << out[3] << "\n";
    int16_t out2[4] = {
        int16_t(-f * (1 << 15)),
        int16_t(-f * INT16_MAX),
        int16_t(-f * (1 << 15) + 0.5),
        int16_t(-f * INT16_MAX + 0.5)
    };
    std::cout << std::setw(5) << -f << ": "
              << std::setw(6) << out2[0] << " "
              << std::setw(6) << out2[1] << " "
              << std::setw(6) << out2[2] << " "
              << std::setw(6) << out2[3] << "\n";
}
```
...Does *NOT*, if you call `x(1.1);`:
```
  1.1: -29492 -29493 -29491 -29492
 -1.1:  29492  29493  29492  29493
```

Apologies for the misinformation. Marking this as "Resolved" since it's not helpful commentary as it turns out.